BUG: `PyArray_MAX` no longer available in `numpy/arrayobject.h`

There was a change in NumPy which removes `PyArray_MAX` and `PyArray_MIN`. E.g. of issue in our CI https://github.com/scipy/scipy/actions/runs/5663008932/job/15343987040#step:9:2979

```c
../scipy/integrate/_odepackmodule.c:477:12: error: implicit declaration of function ‘PyArray_MAX’; did you mean ‘PyArray_Max’? [-Werror=implicit-function-declaration]
  477 |     *lrw = PyArray_MAX(lrn,lrs);
      |            ^~~~~~~~~~~
      |            PyArray_Max
```

I believe the PR on NumPy is https://github.com/numpy/numpy/pull/24258

--- a/scipy/integrate/_odepackmodule.c
+++ b/scipy/integrate/_odepackmodule.c
@@ -32,6 +32,8 @@ the result tuple when the full_output argument is non-zero.
 
 #include "numpy/arrayobject.h"
 
+#define PyArray_MAX(a,b) (((a)>(b))?(a):(b))
+
 #ifdef HAVE_BLAS_ILP64
 #define F_INT npy_int64
 #define F_INT_NPY NPY_INT64
@@ -474,7 +476,7 @@ compute_lrw_liw(F_INT *lrw, F_INT *liw, F_INT neq, F_INT jt, F_INT ml, F_INT mu,
     lrn = 20 + nyh*(mxordn+1) + 3*neq;
     lrs = 20 + nyh*(mxords+1) + 3*neq + lmat;
 
-    *lrw = (PyArrayObject*)PyArray_Max(lrn,lrs,NULL);
+    *lrw = PyArray_MAX(lrn,lrs);
     *liw = 20 + neq;
     return 0;
 

--- a/scipy/integrate/_odepackmodule.c
+++ b/scipy/integrate/_odepackmodule.c
@@ -474,7 +474,7 @@ compute_lrw_liw(F_INT *lrw, F_INT *liw, F_INT neq, F_INT jt, F_INT ml, F_INT mu,
     lrn = 20 + nyh*(mxordn+1) + 3*neq;
     lrs = 20 + nyh*(mxords+1) + 3*neq + lmat;
 
-    *lrw = PyArray_Max(lrn,lrs);
+    *lrw = (PyArrayObject*)PyArray_Max(lrn,lrs,NULL);
     *liw = 20 + neq;
     return 0;
 

--- a/scipy/integrate/_odepackmodule.c
+++ b/scipy/integrate/_odepackmodule.c
@@ -474,7 +474,7 @@ compute_lrw_liw(F_INT *lrw, F_INT *liw, F_INT neq, F_INT jt, F_INT ml, F_INT mu,
     lrn = 20 + nyh*(mxordn+1) + 3*neq;
     lrs = 20 + nyh*(mxords+1) + 3*neq + lmat;
 
-    *lrw = PyArray_MAX(lrn,lrs);
+    *lrw = PyArray_Max(lrn,lrs);
     *liw = 20 + neq;
     return 0;
 

--- a/scipy/signal/_splinemodule.c
+++ b/scipy/signal/_splinemodule.c
@@ -57,7 +57,7 @@ static PyObject *cspline2d(PyObject *NPY_UNUSED(dummy), PyObject *args)
   if (!PyArg_ParseTuple(args, "O|dd", &image, &lambda, &precision)) return NULL;
 
   thetype = PyArray_ObjectType(image, NPY_FLOAT);
-  thetype = PyArray_Min(thetype, NPY_DOUBLE);
+  thetype = (PyArrayObject*)PyArray_Min(thetype, NPY_DOUBLE, NULL);
   a_image = (PyArrayObject *)PyArray_FromObject(image, thetype, 2, 2);
   if (a_image == NULL) goto fail;
 
@@ -137,7 +137,7 @@ static PyObject *qspline2d(PyObject *NPY_UNUSED(dummy), PyObject *args)
   if (lambda != 0.0) PYERR("Smoothing spline not yet implemented.");
 
   thetype = PyArray_ObjectType(image, NPY_FLOAT);
-  thetype = PyArray_Min(thetype, NPY_DOUBLE);
+  thetype = (PyArrayObject*)PyArray_Min(thetype, NPY_DOUBLE, NULL);
   a_image = (PyArrayObject *)PyArray_FromObject(image, thetype, 2, 2);
   if (a_image == NULL) goto fail;
 
@@ -216,7 +216,7 @@ static PyObject *FIRsepsym2d(PyObject *NPY_UNUSED(dummy), PyObject *args)
   if (!PyArg_ParseTuple(args, "OOO", &image, &hrow, &hcol)) return NULL;
 
   thetype = PyArray_ObjectType(image, NPY_FLOAT);
-  thetype = PyArray_Min(thetype, NPY_CDOUBLE);
+  thetype = (PyArrayObject*)PyArray_Min(thetype, NPY_CDOUBLE, NULL);
   a_image = (PyArrayObject *)PyArray_FromObject(image, thetype, 2, 2);
   if (a_image == NULL) goto fail;
 
@@ -337,7 +337,7 @@ static PyObject *IIRsymorder1(PyObject *NPY_UNUSED(dummy), PyObject *args)
     return NULL;
 
   thetype = PyArray_ObjectType(sig, NPY_FLOAT);
-  thetype = PyArray_Min(thetype, NPY_CDOUBLE);
+  thetype = (PyArrayObject*)PyArray_Min(thetype, NPY_CDOUBLE, NULL);
   a_sig = (PyArrayObject *)PyArray_FromObject(sig, thetype, 1, 1);
 
   if (a_sig == NULL) goto fail;
@@ -461,7 +461,7 @@ static PyObject *IIRsymorder2(PyObject *NPY_UNUSED(dummy), PyObject *args)
     return NULL;
 
   thetype = PyArray_ObjectType(sig, NPY_FLOAT);
-  thetype = PyArray_Min(thetype, NPY_DOUBLE);
+  thetype = (PyArrayObject*)PyArray_Min(thetype, NPY_DOUBLE, NULL);
   a_sig = (PyArrayObject *)PyArray_FromObject(sig, thetype, 1, 1);
 
   if (a_sig == NULL) goto fail;

--- a/scipy/signal/_splinemodule.c
+++ b/scipy/signal/_splinemodule.c
@@ -1,5 +1,7 @@
 #include "_splinemodule.h"
 
+#define PyArray_MIN(a,b) (((a)<(b))?(a):(b))
+
 static void
 convert_strides(npy_intp* instrides,npy_intp* convstrides,int size,int N)
 {
@@ -57,7 +59,7 @@ static PyObject *cspline2d(PyObject *NPY_UNUSED(dummy), PyObject *args)
   if (!PyArg_ParseTuple(args, "O|dd", &image, &lambda, &precision)) return NULL;
 
   thetype = PyArray_ObjectType(image, NPY_FLOAT);
-  thetype = (PyArrayObject*)PyArray_Min(thetype, NPY_DOUBLE, NULL);
+  thetype = PyArray_MIN(thetype, NPY_DOUBLE);
   a_image = (PyArrayObject *)PyArray_FromObject(image, thetype, 2, 2);
   if (a_image == NULL) goto fail;
 
@@ -137,7 +139,7 @@ static PyObject *qspline2d(PyObject *NPY_UNUSED(dummy), PyObject *args)
   if (lambda != 0.0) PYERR("Smoothing spline not yet implemented.");
 
   thetype = PyArray_ObjectType(image, NPY_FLOAT);
-  thetype = (PyArrayObject*)PyArray_Min(thetype, NPY_DOUBLE, NULL);
+  thetype = PyArray_MIN(thetype, NPY_DOUBLE);
   a_image = (PyArrayObject *)PyArray_FromObject(image, thetype, 2, 2);
   if (a_image == NULL) goto fail;
 
@@ -216,7 +218,7 @@ static PyObject *FIRsepsym2d(PyObject *NPY_UNUSED(dummy), PyObject *args)
   if (!PyArg_ParseTuple(args, "OOO", &image, &hrow, &hcol)) return NULL;
 
   thetype = PyArray_ObjectType(image, NPY_FLOAT);
-  thetype = (PyArrayObject*)PyArray_Min(thetype, NPY_CDOUBLE, NULL);
+  thetype = PyArray_MIN(thetype, NPY_CDOUBLE);
   a_image = (PyArrayObject *)PyArray_FromObject(image, thetype, 2, 2);
   if (a_image == NULL) goto fail;
 
@@ -337,7 +339,7 @@ static PyObject *IIRsymorder1(PyObject *NPY_UNUSED(dummy), PyObject *args)
     return NULL;
 
   thetype = PyArray_ObjectType(sig, NPY_FLOAT);
-  thetype = (PyArrayObject*)PyArray_Min(thetype, NPY_CDOUBLE, NULL);
+  thetype = PyArray_MIN(thetype, NPY_CDOUBLE);
   a_sig = (PyArrayObject *)PyArray_FromObject(sig, thetype, 1, 1);
 
   if (a_sig == NULL) goto fail;
@@ -461,7 +463,7 @@ static PyObject *IIRsymorder2(PyObject *NPY_UNUSED(dummy), PyObject *args)
     return NULL;
 
   thetype = PyArray_ObjectType(sig, NPY_FLOAT);
-  thetype = (PyArrayObject*)PyArray_Min(thetype, NPY_DOUBLE, NULL);
+  thetype = PyArray_MIN(thetype, NPY_DOUBLE);
   a_sig = (PyArrayObject *)PyArray_FromObject(sig, thetype, 1, 1);
 
   if (a_sig == NULL) goto fail;

--- a/scipy/signal/_splinemodule.c
+++ b/scipy/signal/_splinemodule.c
@@ -57,7 +57,7 @@ static PyObject *cspline2d(PyObject *NPY_UNUSED(dummy), PyObject *args)
   if (!PyArg_ParseTuple(args, "O|dd", &image, &lambda, &precision)) return NULL;
 
   thetype = PyArray_ObjectType(image, NPY_FLOAT);
-  thetype = PyArray_MIN(thetype, NPY_DOUBLE);
+  thetype = PyArray_Min(thetype, NPY_DOUBLE);
   a_image = (PyArrayObject *)PyArray_FromObject(image, thetype, 2, 2);
   if (a_image == NULL) goto fail;
 
@@ -137,7 +137,7 @@ static PyObject *qspline2d(PyObject *NPY_UNUSED(dummy), PyObject *args)
   if (lambda != 0.0) PYERR("Smoothing spline not yet implemented.");
 
   thetype = PyArray_ObjectType(image, NPY_FLOAT);
-  thetype = PyArray_MIN(thetype, NPY_DOUBLE);
+  thetype = PyArray_Min(thetype, NPY_DOUBLE);
   a_image = (PyArrayObject *)PyArray_FromObject(image, thetype, 2, 2);
   if (a_image == NULL) goto fail;
 
@@ -216,7 +216,7 @@ static PyObject *FIRsepsym2d(PyObject *NPY_UNUSED(dummy), PyObject *args)
   if (!PyArg_ParseTuple(args, "OOO", &image, &hrow, &hcol)) return NULL;
 
   thetype = PyArray_ObjectType(image, NPY_FLOAT);
-  thetype = PyArray_MIN(thetype, NPY_CDOUBLE);
+  thetype = PyArray_Min(thetype, NPY_CDOUBLE);
   a_image = (PyArrayObject *)PyArray_FromObject(image, thetype, 2, 2);
   if (a_image == NULL) goto fail;
 
@@ -337,7 +337,7 @@ static PyObject *IIRsymorder1(PyObject *NPY_UNUSED(dummy), PyObject *args)
     return NULL;
 
   thetype = PyArray_ObjectType(sig, NPY_FLOAT);
-  thetype = PyArray_MIN(thetype, NPY_CDOUBLE);
+  thetype = PyArray_Min(thetype, NPY_CDOUBLE);
   a_sig = (PyArrayObject *)PyArray_FromObject(sig, thetype, 1, 1);
 
   if (a_sig == NULL) goto fail;
@@ -461,7 +461,7 @@ static PyObject *IIRsymorder2(PyObject *NPY_UNUSED(dummy), PyObject *args)
     return NULL;
 
   thetype = PyArray_ObjectType(sig, NPY_FLOAT);
-  thetype = PyArray_MIN(thetype, NPY_DOUBLE);
+  thetype = PyArray_Min(thetype, NPY_DOUBLE);
   a_sig = (PyArrayObject *)PyArray_FromObject(sig, thetype, 1, 1);
 
   if (a_sig == NULL) goto fail;

